accord-core/src/main/java/accord/utils/async/AsyncChains.java

```diff
@@ -198,7 +198,14 @@ static class EncapsulatedFlatMap<I, O> extends FlatMap<I, O>
         @Override
         public AsyncChain<O> apply(I i)
         {
-            return map.apply(i);
+            try
+            {
+                return map.apply(i);
+            }
+            catch (Throwable t)
+            {
+                return AsyncChains.failure(t);
```

Comment:
Pushing the error into the chain, does it actually make sense because the error occurred at the construction of the chain not execution?

Reply:
this is the semantic you get with futures, and what we all expect.  Lets look at the following example

```
AsyncChain<?> chain = doLogic();
chain.map(ignore -> {throw new NullPointerException();}.begin(agent);
```

what thread sees the NPE?  *should* agent see it, should this local thread see it, or should the publisher of the data see it?  I strongly feel that `agent` should see it... right now that isn't true

Reply:
Got it, I had gotten confused into thinking that since it returned a `AsyncChain` it wasn't actually executing the body here. Totally agree.